### ISSUE TITLE
Empty `settings.json`

### ISSUE THREAD

**Issue by deanishe** [deanishe/alfred-workflow#69]
After "greenlighting" the killing of AW-based workflows by Alfred, due to atomic, guaranteed writing of data/cache files and `settings.json`, empty `settings.json` files sometimes turn up.

It's not clear whether this is an error in the "atomic write" code or somewhere else.

It's also a very-hard-to-replicate bug.

What are the potential solutions?

**Comment by owenwater** [comment]
I think the writing operation is atomic but not thread safe which may cause the issue. 

With `Terminate previous script` mode, the Alfred kill the old process but I don't think they can guarantee that there is only one process alive at any given time. So the file will be empty with following scenario: 

```
P1 Write the file -> P2 Open the file(clean) -> P2 is terminated -> P1 rename the file -> P1 is terminated.
```

Catching signal turns out a better solution because it makes sure there is something in the file eventually. However, it's still possible that the file is read and written at the same time which raises the exception.

Another workaround is writing something into file iff there is a diff. It's not a solution but just reduce the possibility.

BTW, have you seen this empty `settings.json` issue in other workflow after the code change? If so, how frequently?

**Comment by deanishe** [comment]
> BTW, have you seen this empty settings.json issue in other workflow after the code change? 

Yes. I don't use your calendar workflow. It happened to me with one of my own workflows based on a recent version of AW.

I think your P1 > P2 > P1 idea is the most likely explanation.

The proper solution would appear to be file locking, but I strongly suspect that preventing AW from unconditionally writing the last run version on every run would probably fix the issue in 99% of cases. In particular, this has knock-on issues for other situations: if your workflow creates multiple `Workflow` instances, any changes they make to the settings will be overwritten by the "main" instance (where you called `Workflow.run()` because it write the last run version after the workflow script completes.

In any case, I think AW metadata should probably be kept separate from the workflow data, i.e. in a different file.

Additionally, it'd probably be a good idea for AW to back up `settings.json` before changing it, and perhaps perform a sanity check when loading it (i.e. is it valid JSON).

**Comment by owenwater** [comment]
> Yes. I don't use your calendar workflow. It happened to me with one of my own workflows based on a recent version of AW.

Good, I just want to make sure it's not a bug in my workflow :smile: 

My main problem is that I'm using `setdefault()` to read the dictionary. It causes lots of unnecessary write operation in the end. That's why I propose diff check.

I don't have too much confidence about the backup. It can't solve the issue alone without thread safe. Because the backup file could be messed up just like the temporary file. Without backup, sanity check won't be able to do too much thing. 

Lock is a solution. I can't find any good reason go against it except the performance. Because I just tried my workflow on a 5 years old macbook and everything seems pretty slow already. 

There is another thing I'm thinking about is sqlite as the plate file may just not strong enough at this point. We can at least consider it on V2.

**Comment by deanishe** [comment]
What do you mean by "diff check" exactly?

> My main problem is that I'm using `setdefault()` to read the dictionary.

Instead of calling that multiple times, can't you run `wf.settings.update(default_settings)` once or pass the defaults to the constructor?

> I don't have too much confidence about the backup. It can't solve the issue alone without thread safe. 

It's not meant to solve the issue. It's meant to prevent data loss when there is an issue. By keeping the last 5–10 versions of `settings.json`, if something does go wrong, the user may be able to retrieve their settings from a backup.

**Comment by owenwater** [comment]
> What do you mean by "diff check" exactly?

`diff check` means AW compares the version in the memory and one on the disk before writing and only write when there is a difference.

> Instead of calling that multiple times, can't you run `wf.settings.update(default_settings)` once or pass the defaults to the constructor?

The thing I want to do is reading from the dictionary, adding new entry with default value if it doesn't exist. `update()` or `default_settings` can't handle the case that there is only one entry missing. But I do think about using `default_settings` instead now

>It's not meant to solve the issue. It's meant to prevent data loss when there is an issue. By keeping the last 5–10 versions of `settings.json`, if something does go wrong, the user may be able to retrieve their settings from a backup.

Backup is a good idea. What I'm worry about is that AW is trying to implement its own ACID system. So my question is: Does AW really need this complexity? If so, why don't we just use any existing library.

**Comment by deanishe** [comment]
> Does AW really need this complexity?

It needs to not leave empty `settings.json` files. That's a pretty big bug.

**Comment by deanishe** [comment]
I've added file locking around `settings.json` writes. That should prevent concurrently-running scripts from trampling each other's settings.

**Comment by owenwater** [comment]
Hi @deanishe,

The fix looks good. However, the lock file may not be deleted if workflow is killed while holding the lock. As a result, it will block all following request. I write a [script](https://gist.github.com/owenwater/198b323ecb268d0eaf79) to verify this problem.

You can also use this script to reproduce the empty file issue.

Owen

**Comment by deanishe** [comment]
I'm sorry, I can't follow the script because it requires an external script that isn't included.

Could you provide a fully self-contained demonstration of the issue, so it can go in a unit test?

It should be *relatively* easy to remove stale lock files. The current code writes the PID to the lockfile, and the code from `background.py` can be used to determine if the process is still running or not.

**Comment by owenwater** [comment]
You can just replace [script_name](https://gist.github.com/owenwater/198b323ecb268d0eaf79#file-run-py-L19) to whatever script that write to `settings.json` with AW. Also `script_arg` is the argument that script take.  

If you still can't run the script, just download my [workflow](https://github.com/owenwater/alfred-cal) and put the script in the root directory. You also need to put `self.log.debug('write settings.json start')` and `self.log.debug('write settings.json end')` in [main.py/init_settings()](https://github.com/owenwater/alfred-cal/blob/master/src/main.py#L26) so you can know how many process are killed while update JSON file.

**Comment by deanishe** [comment]
Fixed in #79
